react/ProductAssemblyOptions.tsx

```diff
@@ -0,0 +1,142 @@
+import React, { useMemo, FC, Fragment } from 'react'
+import useProduct from 'vtex.product-context/useProduct'
+import { find, propEq, compose, last, split, pathOr } from 'ramda'
+import ProductAssemblyOptionsGroupStateManager from './components/ProductAssemblyOptions/ProductAssemblyOptionsGroupStateManager'
+
+type PriceMap = Record<string, Record<string, Record<string, number>>>
+
+const findItemMetadata = (id: string) => find<MetadataItem>(propEq('id', id))
+const splitGroupName = compose(
+  last,
+  split('_')
+)
+
+const getItemAssemblyOptions = (
+  selectedItem: SelectedItem,
+  itemMetadata: ItemMetadata
+) => {
+  if (!itemMetadata || !itemMetadata.items) {
+    return null
+  }
+  const metadata = findItemMetadata(selectedItem.itemId)(itemMetadata.items)
+  return metadata && metadata.assemblyOptions
+}
+
+const parsePriceMap = (itemMetadata: ItemMetadata) => {
+  const result = {} as PriceMap
+  for (const priceTableItem of itemMetadata.priceTable) {
+    const { type } = priceTableItem
+    result[type] = {}
+    for (const priceItem of priceTableItem.values) {
+      const groupPrices = result[type][priceItem.assemblyId] || {}
+      groupPrices[priceItem.id] = priceItem.price
+      result[type][priceItem.assemblyId] = groupPrices
+    }
+  }
+  return result
+}
+
+const parseAssemblyOptions = (
+  assemblyOptions: AssemblyOption[],
+  priceMap: PriceMap,
+  product: Product,
+  fatherItemId?: string,
+  treePath?: TreePath[]
+) => {
+  const assemblyOptionsParsed = {} as Record<string, AssemblyOptionGroup>
+  for (const assemblyOption of assemblyOptions) {
+    if (!assemblyOption.composition) {
+      continue
+    }
+    const currentTreePath =
+      fatherItemId && treePath
+        ? [...treePath, { itemId: fatherItemId, groupId: assemblyOption.id }]
+        : []
+    assemblyOptionsParsed[assemblyOption.id] = {
+      id: assemblyOption.id,
+      minQuantity: assemblyOption.composition.minQuantity,
+      maxQuantity: assemblyOption.composition.maxQuantity,
+      items: {},
+      groupName: splitGroupName(assemblyOption.id),
+      treePath: currentTreePath,
+    }
+    const items = {} as Record<string, AssemblyItem>
+    for (const assemblyItem of assemblyOption.composition.items) {
+      const optionMetadata = findItemMetadata(assemblyItem.id)(
+        product.itemMetadata.items
+      )
+
+      // Recursively parse children of this assembly option
+      const children =
+        optionMetadata!.assemblyOptions.length > 0
```

Comment:
why are you asserting `optionMetadata` is not null? are you sure of it?

and at any rate, I think it's generally better to just check if it's null and return (or `break` or `continue` since this is a loop), if nothing else so that typescript knows it has some value and stops complaining

Reply:
yes I am pretty sure it will always find the result